eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -372,5 +372,81 @@ object Sphinx extends Logging {
+   * the end of an onion path. It can be used for payments or onion messages.
+   */
+  object RouteBlinding {
+
+    /**
+     * @param blindedPublicKey     blinded public key, which hides the real public key.
+     * @param blindingEphemeralKey blinding tweak that can be used by the receiving node to derive the private key that
+     *                             matches the blinded public key.
+     * @param encryptedPayload     encrypted payload that can be decrypted with the receiving node's private key and the
+     *                             blinding ephemeral key.
+     */
+    case class BlindedHop(blindedPublicKey: PublicKey, blindingEphemeralKey: PublicKey, encryptedPayload: ByteVector)
+
+    /**
+     * @param introductionNodeId the first node should not be blinded, otherwise the sender cannot locate it.
+     * @param blindedHops        blinded hops.
+     */
+    case class BlindedRoute(introductionNodeId: PublicKey, blindedHops: Seq[BlindedHop]) {
+      val nodeIds: Seq[PublicKey] = introductionNodeId +: blindedHops.tail.map(_.blindedPublicKey)
+    }
+
+    /**
+     * Blind the provided route and encrypt intermediate nodes' payloads.
+     *
+     * @param sessionKey this node's session key.
+     * @param publicKeys public keys of each node on the route, starting from the introduction point.
+     * @param payloads   payloads that should be encrypted for each node on the route.
+     * @return a tuple (blinded public keys, encrypted payloads)
+     */
+    def create(sessionKey: PrivateKey, publicKeys: Seq[PublicKey], payloads: Seq[ByteVector]): BlindedRoute = {
+      require(publicKeys.length == payloads.length, "a payload must be provided for each node in the blinded path")
+      var e = sessionKey
+      val blindedHops = publicKeys.zip(payloads).map { case (publicKey, payload) =>
+        val blindingKey = e.publicKey
+        val sharedSecret = computeSharedSecret(publicKey, e)
+        val blindedPublicKey = blind(publicKey, generateKey("blinded_node_id", sharedSecret))
+        val rho = generateKey("rho", sharedSecret)
+        val (encryptedPayload, mac) = ChaCha20Poly1305.encrypt(rho, zeroes(12), payload, blindingKey.value)
+        e = e.multiply(PrivateKey(Crypto.sha256(blindingKey.value ++ sharedSecret.bytes)))
+        BlindedHop(blindedPublicKey, blindingKey, encryptedPayload ++ mac)
+      }
+      BlindedRoute(publicKeys.head, blindedHops)
+    }
+
+    /**
+     * Compute the blinded private key that must be used to decrypt an incoming blinded onion.
+     *
+     * @param privateKey           this node's private key.
+     * @param blindingEphemeralKey unblinding ephemeral key.
+     * @return this node's blinded private key.
+     */
+    def derivePrivateKey(privateKey: PrivateKey, blindingEphemeralKey: PublicKey): PrivateKey = {
+      val sharedSecret = computeSharedSecret(blindingEphemeralKey, privateKey)
+      privateKey.multiply(PrivateKey(generateKey("blinded_node_id", sharedSecret)))
+    }
```

Comment:
How should this be used? I expected it to be used in the next function(`decryptPayload`) but it's not.

Reply:
See https://github.com/lightningnetwork/lightning-rfc/blob/route-blinding/proposals/route-blinding.md#sending-to-a-blinded-route

When a node is part of a blinded route, the onion it will receive isn't encrypted to its `node_id` so the node cannot decrypt the onion. Instead the onion has been encrypted using the `blinded_node_id`, so the node must be able to derive the corresponding private key *before* decrypting the onion. This function does exactly that.

Reply:
I'll add a test case that describes the complete onion encryption / decryption steps (for the HTLC case), I think it will make it clearer (and it's useful to have).

Reply:
Done in the latest force-push.
I also updated the spec PR, I think you should be able to build upon this for onion messages.